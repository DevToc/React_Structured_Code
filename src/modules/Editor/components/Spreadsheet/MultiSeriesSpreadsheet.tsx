import React, { ReactElement } from 'react';
import { Spreadsheet } from './Spreadsheet';
import { MultiSeriesSpreadsheetProps } from './Spreadsheet.types';

export const MultiSeriesSpreadsheet = (inProps: MultiSeriesSpreadsheetProps): ReactElement => {
  return <Spreadsheet {...inProps} />;
};
