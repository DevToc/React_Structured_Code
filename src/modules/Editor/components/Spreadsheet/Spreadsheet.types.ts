import HotTable from '@handsontable/react';

/**
 * A cell value, which can be anything to support custom cell data types, but by default is `string | number | boolean | undefined`.
 * @see https://github.com/handsontable/handsontable/blob/develop/handsontable/types/common.d.ts
 */
export type CellValue = any;

/**
 * A cell change represented by `[row, column, prevValue, nextValue]`.
 * @see https://github.com/handsontable/handsontable/blob/develop/handsontable/types/common.d.ts
 */
export type CellChange = [number, string | number, CellValue, CellValue];

export interface SpreadsheetProps extends React.ComponentProps<typeof HotTable> {}

export interface SingleSeriesSpreadsheetProps extends SpreadsheetProps {}

export interface MultiSeriesSpreadsheetProps extends SpreadsheetProps {}
