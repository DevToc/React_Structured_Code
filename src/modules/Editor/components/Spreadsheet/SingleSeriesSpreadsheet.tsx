import React, { ReactElement } from 'react';
import { useTranslation } from 'react-i18next';
import { Spreadsheet } from './Spreadsheet';
import { SingleSeriesSpreadsheetProps } from './Spreadsheet.types';

export const SingleSeriesSpreadsheet = (inProps: SingleSeriesSpreadsheetProps): ReactElement => {
  const { t } = useTranslation('editor_spreadsheet', { useSuspense: false });
  const headers = [t('sideMenu.dataTable.category'), t('sideMenu.dataTable.Value')];

  return <Spreadsheet maxCols={2} colHeaders={headers} {...inProps} />;
};
