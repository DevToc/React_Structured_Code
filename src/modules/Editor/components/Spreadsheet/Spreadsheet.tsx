import React, { ReactElement } from 'react';
import { SpreadsheetProps } from './Spreadsheet.types';

// handson table
import 'handsontable/dist/handsontable.full.css';
import { HotTable } from '@handsontable/react';
import { registerAllModules } from 'handsontable/registry';
import { forwardRef } from '@chakra-ui/react';
import { ContextMenu } from 'handsontable/plugins';

// register Handsontable's modules
registerAllModules();

const HANDSON_TABLE_LICENCE = process.env.REACT_APP_HANDSON_TABLE_LICENCE;

export const Spreadsheet = React.memo(
  forwardRef(function Spreadsheet({ data = [], ...otherProps }: SpreadsheetProps, ref): ReactElement {
    return (
      <HotTable
        licenseKey={HANDSON_TABLE_LICENCE}
        id={'hot-spreadsheet'}
        ref={ref}
        data={data}
        colHeaders={true}
        rowHeaders={true}
        width={'auto'}
        height={'auto'}
        undo={false}
        contextMenu={{
          items: {
            row_above: {
              disabled() {
                // Disable add row above when the first row is selected
                return (this.getSelectedLast()?.[0] ?? 0) <= 0;
              },
            },
            row_below: {},
            col_left: {
              disabled() {
                // Disable add col left when the first col is selected
                return (this.getSelectedLast()?.[1] ?? 0) <= 0;
              },
            },
            col_right: {},
            separator_02: ContextMenu.SEPARATOR,
            remove_row: {
              disabled() {
                // Disable remove row when the first row is selected
                return this.getSelectedLast()?.[0] === 0;
              },
            },
            remove_col: {
              disabled() {
                // Disable add col left when the first col is selected
                return (this.getSelectedLast()?.[1] ?? 0) <= 0;
              },
            },
          },
        }}
        {...otherProps}
      />
    );
  }),
);
