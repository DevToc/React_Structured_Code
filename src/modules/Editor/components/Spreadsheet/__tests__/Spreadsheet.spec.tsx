import { renderWithRedux } from '../../../../../utils/test-utils.test';
import { Spreadsheet } from '../Spreadsheet';

describe('Editor/components/Spreadsheet/Spreadsheet.tsx', () => {
  it('should render spreadsheet as expected', () => {
    const testData = [
      {
        id: 1,
        isPromoted: false,
        name: 'Alex',
        score: 10,
        color: 'green',
      },
      {
        id: 2,
        name: 'Adam',
        score: 55,
        isPromoted: false,
      },
      {
        id: 3,
        name: 'Kate',
        score: 61,
        isPromoted: true,
      },
      {
        id: 4,
        name: 'Max',
        score: 98,
        isPromoted: true,
      },
      {
        id: 5,
        name: 'Lucy',
        score: 59,
        isPromoted: false,
      },
    ];

    const { asFragment } = renderWithRedux(<Spreadsheet id={'sheet-1'} data={testData} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
