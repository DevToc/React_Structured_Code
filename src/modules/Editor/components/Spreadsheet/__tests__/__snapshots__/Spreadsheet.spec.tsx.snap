// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor/components/Spreadsheet/Spreadsheet.tsx should render spreadsheet as expected 1`] = `
<DocumentFragment>
  <div
    class="handsontable htRowHeaders htColumnHeaders"
    data-originalstyle="height: auto; overflow: hidden; width: auto;"
    dir=""
    id="sheet-1"
    style="height: auto; overflow: hidden; width: auto;"
  >
    <div
      class="ht_master handsontable"
      dir="ltr"
    >
      <div
        class="wtHolder"
        style="position: relative; height: auto; width: 0px;"
      >
        <div
          class="wtHider"
          style="width: 300px; height: 116px;"
        >
          <div
            class="wtSpreader"
            style="position: relative; top: 0px; left: 0px;"
          >
            <table
              class="htCore"
            >
              <colgroup>
                <col
                  class="rowHeader"
                  style="width: 50px;"
                />
                <col
                  style="width: 50px;"
                />
              </colgroup>
              <thead>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="colHeader cornerHeader"
                      >
                         
                      </span>
                    </div>
                  </th>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="colHeader"
                      >
                        A
                      </span>
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        1
                      </span>
                    </div>
                  </th>
                  <td
                    class=""
                  >
                    1
                  </td>
                </tr>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        2
                      </span>
                    </div>
                  </th>
                  <td
                    class=""
                  >
                    2
                  </td>
                </tr>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        3
                      </span>
                    </div>
                  </th>
                  <td
                    class=""
                  >
                    3
                  </td>
                </tr>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        4
                      </span>
                    </div>
                  </th>
                  <td
                    class=""
                  >
                    4
                  </td>
                </tr>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        5
                      </span>
                    </div>
                  </th>
                  <td
                    class=""
                  >
                    5
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              class="htBorders"
            >
              <div
                style="position: absolute; top: 0px; left: 0px;"
              >
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current corner"
                  style="background-color: rgb(75, 137, 255); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                />
              </div>
              <div
                style="position: absolute; top: 0px; left: 0px;"
              >
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill corner"
                  style="background-color: rgb(255, 0, 0); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ht_clone_top handsontable"
      dir="ltr"
      style="position: absolute; top: 0px; overflow: visible; left: 0px; width: 0px; height: 0px;"
    >
      <div
        class="wtHolder"
        style="position: relative; width: 0px; height: 3px;"
      >
        <div
          class="wtHider"
          style="width: 300px;"
        >
          <div
            class="wtSpreader"
            style="position: relative; left: 0px;"
          >
            <table
              class="htCore"
            >
              <colgroup>
                <col
                  class="rowHeader"
                  style="width: 50px;"
                />
                <col
                  style="width: 50px;"
                />
              </colgroup>
              <thead>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="colHeader cornerHeader"
                      >
                         
                      </span>
                    </div>
                  </th>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="colHeader"
                      >
                        A
                      </span>
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody />
            </table>
            <div
              class="htBorders"
            >
              <div
                style="position: absolute; top: 0px; left: 0px;"
              >
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current corner"
                  style="background-color: rgb(75, 137, 255); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                />
              </div>
              <div
                style="position: absolute; top: 0px; left: 0px;"
              >
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill corner"
                  style="background-color: rgb(255, 0, 0); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ht_clone_bottom handsontable"
      dir="ltr"
      style="position: absolute; top: 0px; overflow: visible; left: 0px;"
    >
      <div
        class="wtHolder"
        style="position: relative;"
      >
        <div
          class="wtHider"
        >
          <div
            class="wtSpreader"
            style="position: relative;"
          >
            <table
              class="htCore"
            >
              <colgroup />
              <thead />
              <tbody />
            </table>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ht_clone_inline_start ht_clone_left handsontable"
      dir="ltr"
      style="position: absolute; top: 0px; overflow: visible; left: 0px; height: 0px; width: 0px;"
    >
      <div
        class="wtHolder"
        style="position: relative; height: 0px; width: 3px;"
      >
        <div
          class="wtHider"
          style="height: 116px;"
        >
          <div
            class="wtSpreader"
            style="position: relative; top: 0px;"
          >
            <table
              class="htCore"
            >
              <colgroup>
                <col
                  class="rowHeader"
                  style="width: 50px;"
                />
              </colgroup>
              <thead>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="colHeader cornerHeader"
                      >
                         
                      </span>
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        1
                      </span>
                    </div>
                  </th>
                </tr>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        2
                      </span>
                    </div>
                  </th>
                </tr>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        3
                      </span>
                    </div>
                  </th>
                </tr>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        4
                      </span>
                    </div>
                  </th>
                </tr>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="rowHeader"
                      >
                        5
                      </span>
                    </div>
                  </th>
                </tr>
              </tbody>
            </table>
            <div
              class="htBorders"
            >
              <div
                style="position: absolute; top: 0px; left: 0px;"
              >
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current corner"
                  style="background-color: rgb(75, 137, 255); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                />
              </div>
              <div
                style="position: absolute; top: 0px; left: 0px;"
              >
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill corner"
                  style="background-color: rgb(255, 0, 0); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ht_clone_top_inline_start_corner ht_clone_top_left_corner handsontable"
      dir="ltr"
      style="position: absolute; top: 0px; overflow: visible; left: 0px; height: 0px; width: 0px;"
    >
      <div
        class="wtHolder"
        style="position: relative;"
      >
        <div
          class="wtHider"
        >
          <div
            class="wtSpreader"
            style="position: relative;"
          >
            <table
              class="htCore"
            >
              <colgroup>
                <col
                  class="rowHeader"
                  style="width: 50px;"
                />
              </colgroup>
              <thead>
                <tr>
                  <th
                    class=""
                  >
                    <div
                      class="relative"
                    >
                      <span
                        class="colHeader cornerHeader"
                      >
                         
                      </span>
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody />
            </table>
            <div
              class="htBorders"
            >
              <div
                style="position: absolute; top: 0px; left: 0px;"
              >
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current"
                  style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                />
                <div
                  class="wtBorder current corner"
                  style="background-color: rgb(75, 137, 255); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                />
              </div>
              <div
                style="position: absolute; top: 0px; left: 0px;"
              >
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill"
                  style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                />
                <div
                  class="wtBorder fill corner"
                  style="background-color: rgb(255, 0, 0); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ht_clone_bottom_inline_start_corner ht_clone_bottom_left_corner handsontable"
      dir="ltr"
      style="position: absolute; overflow: visible; left: 0px; height: 0px; width: 0px;"
    >
      <div
        class="wtHolder"
        style="position: relative;"
      >
        <div
          class="wtHider"
        >
          <div
            class="wtSpreader"
            style="position: relative;"
          >
            <table
              class="htCore"
            >
              <colgroup />
              <thead />
              <tbody />
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
